.github/workflows/docker-publish.yml

```diff
@@ -0,0 +1,198 @@
+          load: true
+          target: builder
+
+      - name: Run unit tests
+        run: |
+          docker-compose -f docker/unit-tests.test.yml \
+            up --exit-code-from=sut
+        env:
+          IMAGE: ${{ env.TEST_TAG }}:builder
+
+  recipe-test:
+    name: Run recipe tests
+    needs: [build]
+    runs-on: ubuntu-latest
+    steps:
+      - name: Repository checkout
+        uses: actions/checkout@v2
+
+      - name: Setup Docker buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Use cache
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.build-cache
+          key: ${{ runner.os }}-docker-${{ github.sha }}
+
+      - name: Build Docker image
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          push: false
+          tags: ${{ env.TEST_TAG }}
+          file: docker/Dockerfile
+          cache-from: type=local,src=/tmp/.build-cache
+          load: true
+
+      - name: Run test recipes on host
+        run: |
+          docker-compose -f docker/recipes.test.yml \
+            up --exit-code-from=sut
+        env:
+          IMAGE: ${{ env.TEST_TAG }}
+
+  uml-test:
+    name: Run UML tests
+    needs: [build]
+    runs-on: ubuntu-latest
+    steps:
+      - name: Repository checkout
+        uses: actions/checkout@v2
+
+      - name: Setup Docker buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Use cache
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.build-cache
+          key: ${{ runner.os }}-docker-${{ github.sha }}
+
+      - name: Build Docker image
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          push: false
+          tags: ${{ env.TEST_TAG }}
+          file: docker/Dockerfile
+          cache-from: type=local,src=/tmp/.build-cache
+          load: true
+
+      - name: Run test recipes using UML backend
+        run: |
+          docker-compose -f docker/recipes-test-uml.yml \
+            up --exit-code-from=sut
+        env:
+          IMAGE: ${{ env.TEST_TAG }}
+
+  publish:
+    name: Publish
+    needs:
+      - unit-test
+      - recipe-test
+      - uml-test
+    if: github.event_name != 'pull_request'
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+      packages: write
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+
+      - name: Extract Docker metadata
+        id: meta
+        uses: docker/metadata-action@v3
+        with:
+          images: ${{ env.DOCKERHUB_TAG }},${{ env.GITHUB_TAG }}
+
+      - name: Login to GitHub Container Registry
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
```

Comment:
we will need to add `DOCKERHUB_USERNAME` and `DOCKERHUB_PASSWORD` as secrets and make sure they cannot be accessed by other pipeline triggerers :-)